### DISCUSSION TITLE
Restore audio and subtitle streams when switching to the previous track

### DISCUSSION THREAD

**Discussion by dmitrylyzo** [jellyfin/jellyfin-web#6139] [General]
> > @dmitrylyzo Hey, sorry, I can't seem to find `isPreviousTrack` anywhere, what exactly are you referring to?
> 
> It doesn't exist. We need to introduce it here:
> 
> https://github.com/jellyfin/jellyfin-web/blob/1114b19b00db6616827c3d2148e21aeeb2f90bfb/src/components/playback/playbackmanager.js#L3076
> 
> Then, we can use it to skip "autoset".

Apologies if this is not the right place to continue this discussion.

Skipping autoset when going back doesn't seem to always pick the previously selected tracks in my testing. I've tried it like this:

1. Start Episode 1
2. Select second audio track
3. Go to Episode 2
4. Select the first (default) audio track
5. Go back to Episode 1
6. First audio track is selected (confirmed autoset did not run)

Manually POSTing to PlaybackInfo also does not seem to return the previously specified stream indexes when omitted. 

This is my branch https://github.com/viown/jellyfin-web/commits/keep-settings-on-previous-track/

_Originally posted by @viown in https://github.com/jellyfin/jellyfin-web/issues/6112#issuecomment-2382642581_

**Comment by dmitrylyzo** [comment]
It looks like we need 2 things:
1. Save selected streams during playback - we send `Progress` and `Stopped`.
2. Read saved streams on the server side if the user has the corresponding options enabled.
https://github.com/jellyfin/jellyfin/blob/9ef7ccfc125d00c0a0ce789626db258a4cb51f2f/MediaBrowser.Model/Dlna/StreamBuilder.cs#L648
https://github.com/jellyfin/jellyfin/blob/9ef7ccfc125d00c0a0ce789626db258a4cb51f2f/MediaBrowser.Model/Dlna/StreamBuilder.cs#L642
Or get `MediaSourceInfo` with `DefaultAudioStreamIndex` and `DefaultSubtitleStreamIndex` on the client side and use them: https://github.com/jellyfin/jellyfin-web/blob/0728bad2229931c611c23ac40ed751c5827fae7c/src/components/playback/playbackmanager.js#L2577
But last time I tested, they were missing.

**Reply by viown** [reply]
Okay, so it looks like progress is being sent and the indexes get updated, but it's because I was moving episodes too fast (before the progress report got sent) and that's why it didn't work in my testing?

**Reply by viown** [reply]
Nevermind, this still does not seem to always be the case when the indexes in options are filled in. So when the track is a previous track, should we also set DefaultAudioStreamIndex and DefaultSubtitleStreamIndex from media source?

**Reply by dmitrylyzo** [reply]
> So when the track is a previous track, should we also set DefaultAudioStreamIndex and DefaultSubtitleStreamIndex from media source?

If we are going to fix this on the client side: we should set `options.audioStreamIndex` and `options.subtitleStreamIndex` from media source, but `Default*` fields were not populated in [`getMediaStreams`](https://github.com/jellyfin/jellyfin-web/blob/0728bad2229931c611c23ac40ed751c5827fae7c/src/components/playback/playbackmanager.js#L2577) when I tested.

On the other hand, we should probably always restore saved streams on the server ([1](https://github.com/jellyfin/jellyfin/blob/9ef7ccfc125d00c0a0ce789626db258a4cb51f2f/MediaBrowser.Model/Dlna/StreamBuilder.cs#L648), [2](https://github.com/jellyfin/jellyfin/blob/9ef7ccfc125d00c0a0ce789626db258a4cb51f2f/MediaBrowser.Model/Dlna/StreamBuilder.cs#L642)) if they are not specified in the request (_currently, it applies auto selection_). But we also need to be able to reset them so the server applies auto selection.
